experimental work on EC operations optimizations

ltvvlz opened the pull request:
This pull request implements:

1.Batched Affine point addition. This functionality allows the optimized addition of points in an array of vectors of group elements such that elements of each vector of the array are all summed up into a single group element. Thus, the array of vectors of group elements is summed up into the array of group elements which are the sums of elements of the corresponding vectors. The Affine addition field element inversions are all batched across the complete array of vectors in rounds yielding the "logarithmic" (with respect to the number of resulting inversions) performance.
2. Implementation of the Pippenger multiexponentiation algorithm with the use of the optimized Affine addition.


Use:
     to exercise the implemented functionality, run the tests and benchmarks as this:
     cargo test --release -- --nocapture --test-threads 1 variable_base


Observations with the current implementation:

1. Batched Affine point addition algorithm provides on average, as evidenced by the attached benchmark screenshot, ~35% performance improvement over the use of the serial mixed Jacobian addition.

2. Use of the optimized Affine addition for the multiexponentiation Pippernger algorithm proves, as evidenced by the attached benchmark screenshot, to produce a negative performance effect on the multiexponentiation functionality. The positive effects of the optimised Affine addition are negated by the use of memory allocations and memory IO. Performance improvement of the optimized Affine addition relies on the array of vectors of group elements to be in advance prepared for the summation. Pippenger (in this form), on the other hand, has to additionally deal with the allocation of the arrays of vectors (in parallel under multithreading) and the movement of elements into its vectors. These memory allocations and memory IO produce the negative performance effect. Thanslating to the Pippenger lexicon, the arrays of vectors correspond to the Windows whereas the vectors themselves correspond to Buckets.
![Screenshot from 2020-04-26 13-12-29](https://user-images.githubusercontent.com/59975641/80314478-b695df00-87bf-11ea-84a9-01261e358d24.png)

![Affine-batched-addition-performance](https://user-images.githubusercontent.com/59975641/80314592-72570e80-87c0-11ea-9642-3a2e4424b845.png)
![multiexp-with-affine-bathed-addition-performance](https://user-images.githubusercontent.com/59975641/80314595-771bc280-87c0-11ea-9639-a4a47777a9d1.png)

ltvvlz committed:
experimental work on EC operations optimizations

imeckler commented:
Hm -- that's a shame.. where do you think we go from here? Is it possible that optimizing the memory usage (e.g., by just reusing memory during the affine "reduction") will help us take advantage of the batched affine addition?

And is it the case that e.g., the batched affine pippenger is still superior in the single threaded case?

imeckler commented:
One more thought -- maybe we should iterate across windows serially and do the parallelism across the reduction (i.e., within a single window) since that should give us better memory locality

ltvvlz committed:
This commit attempts to to improve the multiexp Pippenger algorithm performance with the use of batched Affine point addition with changing the threadi ng across the windows to the threading across the buckets of the point data. This change does positively affects the memory usage since only a single allocation of memory for the buckets of a window is needed. However, this improvement appears to still negatively affect, in a lesses degree, the the algorithm performance such that the benefits of the Affine batch addition are lost.

The peek of the improvemance was tested on multiexp array sizes of > 100000. At this peek, the performance is still ~35% inferior to that of the algorithm using serial Jacobian mixed addition. For the multiexp arrays arrays of sizes smaller than 10000 the change appears to have the negative effect.

*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 20.170312495s
     serial: 14.647667983s
     batch/serial: 1.3770324
Length: 100000
     batch: 2.589379652s
     serial: 1.979924237s
     batch/serial: 1.3078176
Length: 10000
     batch: 451.397416ms
     serial: 264.167974ms
     batch/serial: 1.7087516
Length: 1000
     batch: 133.422168ms
     serial: 47.053773ms
     batch/serial: 2.8355253
Length: 100
     batch: 47.473296ms
     serial: 11.148152ms
     batch/serial: 4.2584004
Length: 10
     batch: 20.414964ms
     serial: 3.451074ms
     batch/serial: 5.9155393
Length: 1
     batch: 3.50487ms
     serial: 1.965183ms
     batch/serial: 1.7834827
ok

ltvvlz committed:
This commits the flattened version of the batch Affine addition function where the passed vector of vectors (window of buckets) is flattened into a single flat vector. Benchmark msm::variable_base::test::batch_addition shows the following degradation of performance with larger number of smaller buckets also with the flattened structure:

running 1 test
test msm::variable_base::test::batch_addition ...
*****BENCHMARKING OPTIMISED AFFINE VS SERIAL JACOBIAN MIXED ADDITION*****

Length: 1000000
Buckets: 1
     serial: 338.0628ms
     batch: 218.546781ms
     batch/serial: 0.646468

Length: 100000
Buckets: 10
     serial: 945.785013ms
     batch: 618.299604ms
     batch/serial: 0.6537422

Length: 10000
Buckets: 100
     serial: 868.435418ms
     batch: 690.321573ms
     batch/serial: 0.7949027

Length: 1000
Buckets: 1000
     serial: 954.944723ms
     batch: 1.60738435s
     batch/serial: 1.6832224

Length: 100
Buckets: 10000
     serial: 1.171245414s
     batch: 10.101511688s
     batch/serial: 8.62459

Length: 10
Buckets: 100000
     serial: 3.370445821s
     batch: 67.313590532s
     batch/serial: 19.971718
ok

ltvvlz committed:
This commit parallelises the multiexp algorithm with allocating a single window bucket memory and threading across subslices of the list of buckets. The number of subslices is divisible by the number of CPU cores to achieve the parallelism. It appears that with this threading method, the optimal parallel CPU utilization is hindered by accesses to memory whith many context switches. Average CPU utilisation is at about 50%. On the contrary, upon the execution of the serial mixed addition, the CPU utilization is clean 100% across all CPU cores. This difficulty with the optimal CPU utilization makes overal multiex algorithm slower.

Benchmark data for this commit is the following:

running 3 tests
test msm::variable_base::test::batch_addition ...
*****BENCHMARKING OPTIMISED AFFINE VS SERIAL JACOBIAN MIXED ADDITION*****

Length: 1000000
Buckets: 1
     serial: 883.0619ms
     batch: 588.142108ms
     batch/serial: 0.66602594

Length: 100000
Buckets: 10
     serial: 1.010428598s
     batch: 641.376597ms
     batch/serial: 0.634757

Length: 10000
Buckets: 100
     serial: 822.03669ms
     batch: 523.627568ms
     batch/serial: 0.6369881

Length: 1000
Buckets: 1000
     serial: 959.905889ms
     batch: 594.796909ms
     batch/serial: 0.6196409

Length: 100
Buckets: 10000
     serial: 1.1836002s
     batch: 563.725951ms
     batch/serial: 0.47628072

Length: 10
Buckets: 100000
     serial: 3.291431114s
     batch: 489.21484ms
     batch/serial: 0.14863287
ok
test msm::variable_base::test::multiexp ...
*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 18.390284423s
     serial: 14.010344834s
     batch/serial: 1.3126218
Length: 100000
     batch: 3.076862021s
     serial: 1.906286611s
     batch/serial: 1.6140606
Length: 10000
     batch: 521.45769ms
     serial: 262.170714ms
     batch/serial: 1.9890003
Length: 1000
     batch: 145.637526ms
     serial: 44.014236ms
     batch/serial: 3.3088732
Length: 100
     batch: 45.385723ms
     serial: 9.617103ms
     batch/serial: 4.719272
Length: 10
     batch: 19.2247ms
     serial: 3.532978ms
     batch/serial: 5.4414997
Length: 1
     batch: 4.588652ms
     serial: 2.561215ms
     batch/serial: 1.7915918
ok
test msm::variable_base::test::test_with_bn_382_unequal_numbers ... ok

ltvvlz committed:
This commit,unlike the previous, parallelises the multiexp algorithm with allocating a window bucket memory for every thread and threading across windows. TThis method shows higher memory usage but also better ovaral CPU utilisation than the method of the previous commit.

Multiexp benchmark data for this commit is the following:

running 1 test
test msm::variable_base::test::multiexp ...
*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 14.714780079s
     serial: 14.083463807s
     batch/serial: 1.0448267
Length: 100000
     batch: 1.501193996s
     serial: 1.912737335s
     batch/serial: 0.78484064
Length: 10000
     batch: 197.546221ms
     serial: 257.758997ms
     batch/serial: 0.76639897
Length: 1000
     batch: 35.687238ms
     serial: 43.740071ms
     batch/serial: 0.81589353
Length: 100
     batch: 10.428082ms
     serial: 8.736375ms
     batch/serial: 1.1936395
Length: 10
     batch: 3.485928ms
     serial: 2.948079ms
     batch/serial: 1.1824405
Length: 1
     batch: 1.183998ms
     serial: 1.893741ms
     batch/serial: 0.6252164
ok

ltvvlz committed:
Choice of the window size noticeably affects the benefit of the use of the optimised Affine addition in the multiexp algorithm. This commit completes the following:

1. Adjusted the computation of the window size that appears to result in more optimal benefit of the use of the optimised Affine addition in the multiexp algorithm.

2. Removes the flattening of the [Vec<Self>] slice of vectors from the AffineCurve::add_points function. The reason for the removal is the fact that the flattening does not appear to positively affect the performance. Possible reasons are: a. flattening itself consumes CPU dragging performance down, b. flattening does not eliminate the second level of indirection of using vector of vectors (pointers to pointers) since it substitutes it by the indirection of the use of indexes into the offsets into the flattened vector.

To run the benchmarks, use the following:
cargo test --release -- --nocapture --test-threads 1 variable_base

Benchmark results for this commit:

test msm::variable_base::test::batch_addition ...
*****BENCHMARKING OPTIMISED AFFINE VS SERIAL JACOBIAN MIXED ADDITION*****

Length: 1000000
Buckets: 1
     serial: 552.737333ms
     batch: 347.319349ms
     batch/serial: 0.6283624

Length: 100000
Buckets: 10
     serial: 774.365714ms
     batch: 481.364967ms
     batch/serial: 0.6216248

Length: 10000
Buckets: 100
     serial: 879.833883ms
     batch: 546.373258ms
     batch/serial: 0.62099594

Length: 1000
Buckets: 1000
     serial: 922.771868ms
     batch: 549.468118ms
     batch/serial: 0.5954539

Length: 100
Buckets: 10000
     serial: 1.146993667s
     batch: 544.679024ms
     batch/serial: 0.4748754

Length: 10
Buckets: 100000
     serial: 3.293458089s
     batch: 457.050354ms
     batch/serial: 0.13877521
ok
test msm::variable_base::test::multiexp ...
*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 11.8200979s
     serial: 13.19562328s
     batch/serial: 0.895759
Length: 100000
     batch: 1.280504612s
     serial: 1.791433309s
     batch/serial: 0.7147933
Length: 10000
     batch: 178.373434ms
     serial: 255.188368ms
     batch/serial: 0.6989873
Length: 1000
     batch: 30.057456ms
     serial: 40.818154ms
     batch/serial: 0.73637474
Length: 100
     batch: 9.585171ms
     serial: 9.152239ms
     batch/serial: 1.0473033
Length: 10
     batch: 3.529125ms
     serial: 2.948354ms
     batch/serial: 1.1969814
Length: 1
     batch: 1.164914ms
     serial: 1.936549ms
     batch/serial: 0.6015412
ok

ltvvlz committed:
Minor optimizations with the following benchmark results:

running 3 tests
test msm::variable_base::test::batch_addition ...
*****BENCHMARKING OPTIMISED AFFINE VS SERIAL JACOBIAN MIXED ADDITION*****

Length: 1000000
Buckets: 1
     serial: 639.865917ms
     batch: 411.684307ms
     batch/serial: 0.64339155

Length: 100000
Buckets: 10
     serial: 1.379444686s
     batch: 853.498763ms
     batch/serial: 0.6187264

Length: 10000
Buckets: 100
     serial: 1.352976738s
     batch: 839.743768ms
     batch/serial: 0.6206638

Length: 1000
Buckets: 1000
     serial: 1.484178665s
     batch: 885.478075ms
     batch/serial: 0.5966115

Length: 100
Buckets: 10000
     serial: 1.899015346s
     batch: 875.483301ms
     batch/serial: 0.4610196

Length: 10
Buckets: 100000
     serial: 5.282255207s
     batch: 730.329491ms
     batch/serial: 0.13826092
ok
test msm::variable_base::test::multiexp ...
*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 18.472264998s
     serial: 21.380674475s
     batch/serial: 0.86397016
Length: 100000
     batch: 1.971185807s
     serial: 2.83228135s
     batch/serial: 0.6959711
Length: 10000
     batch: 279.988628ms
     serial: 392.799534ms
     batch/serial: 0.7128029
Length: 1000
     batch: 46.259117ms
     serial: 67.502753ms
     batch/serial: 0.6852923
Length: 100
     batch: 13.953221ms
     serial: 16.62628ms
     batch/serial: 0.8392269
Length: 10
     batch: 5.598577ms
     serial: 5.129839ms
     batch/serial: 1.0913749
Length: 1
     batch: 1.997877ms
     serial: 2.767078ms
     batch/serial: 0.7220168
ok

ltvvlz commented:
Latest commit to the branch results in the following benefit as evidenced by the benchmarks:

running 3 tests
test msm::variable_base::test::batch_addition ... 
*****BENCHMARKING OPTIMISED AFFINE VS SERIAL JACOBIAN MIXED ADDITION*****

Length: 1000000
Buckets: 1
     serial: 639.865917ms
     batch: 411.684307ms
     batch/serial: 0.64339155

Length: 100000
Buckets: 10
     serial: 1.379444686s
     batch: 853.498763ms
     batch/serial: 0.6187264

Length: 10000
Buckets: 100
     serial: 1.352976738s
     batch: 839.743768ms
     batch/serial: 0.6206638

Length: 1000
Buckets: 1000
     serial: 1.484178665s
     batch: 885.478075ms
     batch/serial: 0.5966115

Length: 100
Buckets: 10000
     serial: 1.899015346s
     batch: 875.483301ms
     batch/serial: 0.4610196

Length: 10
Buckets: 100000
     serial: 5.282255207s
     batch: 730.329491ms
     batch/serial: 0.13826092
ok
test msm::variable_base::test::multiexp ... 
*****BENCHMARKING MULTIEXP WITH OPTIMISED AFFINE VS*****
******MULTIEXP WITH SERIAL JACOBIAN MIXED ADDITION******
Length: 1000000
     batch: 18.472264998s
     serial: 21.380674475s
     batch/serial: 0.86397016
Length: 100000
     batch: 1.971185807s
     serial: 2.83228135s
     batch/serial: 0.6959711
Length: 10000
     batch: 279.988628ms
     serial: 392.799534ms
     batch/serial: 0.7128029
Length: 1000
     batch: 46.259117ms
     serial: 67.502753ms
     batch/serial: 0.6852923
Length: 100
     batch: 13.953221ms
     serial: 16.62628ms
     batch/serial: 0.8392269
Length: 10
     batch: 5.598577ms
     serial: 5.129839ms
     batch/serial: 1.0913749
Length: 1
     batch: 1.997877ms
     serial: 2.767078ms
     batch/serial: 0.7220168
ok
